engine/collection/compliance/core.go

```diff
@@ -330,8 +337,7 @@ func (c *Core) processBlockProposal(proposal *cluster.Block) error {
 	if err != nil {
 		if state.IsInvalidExtensionError(err) {
 			// if the block proposes an invalid extension of the cluster state, then the block is invalid
-			// TODO: we should slash the block proposer
-			return engine.NewInvalidInputErrorf("invalid extension of cluster state (block: %x, height: %d): %w", blockID, header.Height, err)
+			return model.NewInvalidProposalErrorf(hotstuffProposal, "invalid extension of cluster state (block: %x, height: %d): %w", blockID, header.Height, err)
```

Comment:
If the payload part is invalid (something not reflected in `hotstuff.Proposal` type) then we will lose the information required to produce a slashing challenge. Using the hotstuff error here is arguably a bit misleading, because from HotStuff's perspective the proposal (header) might be completely valid -- it could be only an invalid payload. 

For the time being, this is sufficient for tracking and reporting who committed a violation, but we will eventually need to use a different error type here to later construct a complete slashing challenge.

Reply:
I agree, eventually I think we will use different methods to differentiate on types of slashable scenarios, we could also consider generating proofs for slashing right in validation code and reporting them using consumers. One thing for sure, this part will be changed in future.